ext/js/dom/dom-text-scanner.js

```diff
@@ -177,54 +177,54 @@ export class DOMTextScanner {
         let remainder = this._remainder;
         let newlines = this._newlines;
 
-        while (offset < nodeValueLength) {
+        probeForward: while (offset < nodeValueLength) {
             const char = StringUtil.readCodePointsForward(nodeValue, offset, 1);
             offset += char.length;
             const charAttributes = DOMTextScanner.getCharacterAttributes(char, preserveNewlines, preserveWhitespace);
 
-            if (charAttributes === 0) {
-                // Character should be ignored
-                continue;
-            } else if (charAttributes === 1) {
-                // Character is collapsible whitespace
-                lineHasWhitespace = true;
-            } else {
-                // Character should be added to the content
-                if (newlines > 0) {
-                    if (content.length > 0) {
-                        const useNewlineCount = Math.min(remainder, newlines);
-                        content += '\n'.repeat(useNewlineCount);
-                        remainder -= useNewlineCount;
-                        newlines -= useNewlineCount;
-                    } else {
-                        newlines = 0;
-                    }
-                    lineHasContent = false;
-                    lineHasWhitespace = false;
-                    if (remainder <= 0) {
-                        offset -= char.length; // Revert character offset
-                        break;
-                    }
-                }
-
-                lineHasContent = (charAttributes === 2); // 3 = character is a newline
-
-                if (lineHasWhitespace) {
-                    if (lineHasContent) {
-                        content += ' ';
+            matchCharAttributes: switch (charAttributes) {
```

Comment:
Hmm, personally I don't really like switches in JS because of their complex control flow and it being easy to accidentally introduce bugs when you miss a break etc.
(In contrast I love pattern matching in functional languages, as well as match in Rust, because it is clear how the type is getting broken down without the added complex control flow.)
I'd be curious what @toasted-nutbread thinks, should we discourage or encourage switch statements over ifs?

Reply:
Switches aren't a problem, I would recommend them over the alternative of if-else chains on the same variable. However, I would recommend against labelled breaks, which is what is being used here, as that makes the control flow unclear, and usually indicates that a function should be split up into multiple sub-functions.

Reply:
Eslint should also prevent missed breaks.

Reply:
Switch statements in theory should have better perf compared to if-else chains as well.

Reply:
> Switches aren't a problem, I would recommend them over the alternative of if-else chains on the same variable. However, I would recommend against labelled breaks, which is what is being used here, as that makes the control flow unclear, and usually indicates that a function should be split up into multiple sub-functions.

I also had this in mind when I was writing this. I will seperate them into inner functions.